### PR TITLE
Implement cross-report privacy loss optimization with attributionObject API

### LINKED ISSUE #94: Implement cross-report privacy loss optimization with `attributionObject` API
We currently only have a simplified, non-adaptive, implementation where all the intermediaries have to specify their parameters upfront. It would be nice to support the adaptive API described in the paper and analyzed in the appendix, and simplify intermediary handling along the way.

### PR DESCRIPTION
**Pull request by giorgi-o**
This PR implements the `attributionObject` API under the `experimental` feature flag. It also removes the old cross-report optimization code from the non-experimental pdslib code.

Changes:
- Removed the `EventUris.intermediary_uris` and `ReportRequestUris.intermediary_uris` fields
- `request.compute_report()` and `pds.compute_report()` returns only one report for the querier, instead of a report per querier/intermediary
- Implemented the `attributionObject` API in cross_report.rs
- Make `HistogramRequest` a sub-trait of `EpochReportRequest`, and simplify its associated types
- Delete the `QueryComputeResult<>` struct, as we never return multiple reports simultaneously for one query

Fixes #94

### CODE REVIEW

File: src/pds/cross_report.rs [hunk 1]
```diff
@@ -0,0 +1,404 @@
...
+
+            // Do not consume per-querier, that is done in get_report().
+            // Change the per-querier budgets to 0 to still initialize the
+            // filter.
+            for querier_uri in &uris.querier_uris {
+                filters_to_consume.insert(
+                    FilterId::PerQuerier(epoch_id, querier_uri.clone()),
+                    &0.0, // no budget consumption for per-querier filter
+                );
+            }
```
Comment (tholop) [inline src/pds/cross_report.rs:177]: What happens if you drop this and don't initialize the filters?

File: src/pds/cross_report.rs [hunk 2]
```diff
@@ -0,0 +1,404 @@
...
+        for epoch_id in epochs {
+            let individual_privacy_loss =
+                request.histogram_multi_epoch_report_global_sensitivity();
+
+            let source_losses = compute_epoch_source_losses(
+                &request,
+                relevant_events.sources_for_epoch(&epoch_id),
+                &unfiltered_result,
+                num_epochs,
+            );
```
Comment (tholop) [inline src/pds/cross_report.rs:158]: In the pseudocode I'm using 2a^max for source losses too (eps^i), instead of the tighter individual budget.

File: src/pds/cross_report.rs [hunk 3]
```diff
@@ -0,0 +1,404 @@
...
+    ERR: From<FS::Error>,
+{
+    pub fn measure_conversion(
+        &mut self,
+        request: PpaHistogramRequest<U>,
+        mut relevant_events: RelevantEvents<PpaEvent<U>>,
+    ) -> Result<AttributionObject<PpaHistogramRequest<U>>, ERR> {
+        let uris = request.report_uris();
+
+        let epochs = request.epoch_ids();
```
Comment (tholop) [inline src/pds/cross_report.rs:142]: Could you add something as follows, if epoch_ids has more than two elements?
```
warn!("Cross-report optimization only saves budget when requesting more than 1 epoch. We recommend using the regular API otherwise.")
```

File: src/pds/cross_report.rs [hunk 4]
```diff
@@ -0,0 +1,404 @@
...
+                relevant_events.drop_epoch(&epoch_id);
+
+                // Keep track of why we dropped this epoch
+                oob_filters.push(filter_id);
+            }
+        }
+
+        // Now that we've dropped OOB epochs, we can compute the final report.
+        let mut filtered_report = self.request.compute_report(&relevant_events);
+
```
Comment (tholop) [inline src/pds/cross_report.rs:111]: So here I'm worried `compute_report` will run only on the events that were not requested before, potentially reattributing the value to a different event in each `get_report` call (e.g., if we are doing last touch attribution). 

That's why in the pseudocode I am "freezing" the attribution function a_F, i.e., the mapping from event to values. Whereas `compute_report` recomputes a new mapping each time. 

To implement this, I think you'll have to open up the `compute_report` black box and look at how it is implemented specifically for histogram requests (or even PpaHistogramRequests)

**Review by tholop — CHANGES_REQUESTED** [review]
Thanks @giorgi-o, the shape of the API is great, and it's nice to clean these other things too. There is only one big change, which is that `measure_conversion` should attribute the value to events only once, and then store it, to have consistent attribution across `get_reports` call. Right now, `compute_report` re-attributes the value each time, if I understand correctly. Actually, it's something we could add a test for! Try to create a situation where we have two epochs and some last touch attribution logic, and check that the total attributed value is below the `max_attributable_value`.

File: src/pds/private_data_service.rs [hunk 5]
```diff
@@ -41,6 +43,17 @@ pub struct PdsReport<Q: EpochReportRequest> {
...
+impl<Q: EpochReportRequest> Default for PdsReport<Q> {
+    fn default() -> Self {
+        Self {
+            filtered_report: Q::Report::default(),
+            unfiltered_report: Q::Report::default(),
+            oob_filters: Vec::new(),
+        }
+    }
+}
+
```
[the remark this answers is not in the document]
Comment (giorgi-o) [reply src/pds/private_data_service.rs:56]: The issue is that `#[derive(Default)]` expands to: (shortened)

```rust
impl<Q: EpochReportRequest + Default> Default for PdsReport<Q>
where
    Q::Report: Default,
    Q::EpochId: Default, // <- here
    Q::Uri: Default, // <- and here
{
    // ...
}
```

Meaning that if `Q::EpochId` or `Q::Uri` don't implement `Default` (which they don't), then this does nothing. Unfortunately the #derive isn't smart enough to realise that, so we need to do the impl manually.

File: src/pds/cross_report.rs (hunk 1 above)
Comment (giorgi-o) [reply src/pds/cross_report.rs:177]: Nothing if we assume static capacities. But if the default per-site capacity changes between now and `get_report()`, then I'm guessing the privacy proofs don't apply anymore

File: src/pds/cross_report.rs (hunk 4 above)
Comment (giorgi-o) [reply src/pds/cross_report.rs:111]: I'm not sure that happens. If events were previously requested, they are not removed from `relevant_events`. So `compute_report()` has all the events when it calculates the attribution.

However that raises the question of, what about out-of-budget epochs? Say there are 2 events, one in epoch 5 and one in epoch 6. Say they are both identical (except for the epoch), and so they are both relevant to this report request. There are 2 things I'm not sure about:
1. Under last-touch, does the second event shadow the first? Do they both deduct per-querier budget, or only the second one?
2. Say epoch 6 is out-of-budget. This implementation would remove all events in epoch 6, thus making 5 the most recent relevant event from the point of view of a_F. That sounds wrong, right? How would out-of-budget epochs be handled?
3. Since we freeze a_F, would it be correct to calculate the whole histogram in `measure_conversion()`, and then just return the requested buckets from that histogram on each call to `get_report()`?

File: src/pds/core.rs [hunk 6]
```diff
@@ -150,25 +144,15 @@ where
...
+        let filtered_report = request.compute_report(&relevant_events);
+        debug!("Filtered report: {filtered_report:?}");
 
         let main_report = PdsReport {
-            filtered_report: filtered_result
-                .uri_report_map
-                .get(querier_uri)
-                .unwrap()
-                .clone(),
+            filtered_report,
```
[the remark this answers is not in the document]
Comment (giorgi-o) [reply src/pds/core.rs:152]: Fixed in https://github.com/columbia/pdslib/pull/95/commits/a838df9f69e4818b211d389f7ea0932cf34ad254 and https://github.com/columbia/pdslib/pull/95/commits/709930b2dc17ad9e371724042c01faff80eba0a8

File: src/queries/ppa_histogram.rs [hunk 7]
```diff
@@ -12,20 +12,20 @@ use crate::{
...
-type PpaBucketKey = u64;
-type PpaEpochId = u64;
+pub type PpaBucketKey = u64;
+pub type PpaEpochId = u64;
 
 pub struct PpaRelevantEventSelector<U: Uri = String> {
     pub report_request_uris: ReportRequestUris<U>,
-    pub is_matching_event: Box<dyn Fn(u64) -> bool>,
-    pub bucket_intermediary_mapping: HashMap<u64, U>,
+    pub is_matching_event: Box<dyn Fn(u64 /* filter_data */) -> bool>,
```
[the remark this answers is not in the document]
Comment (giorgi-o) [reply src/queries/ppa_histogram.rs]: Done in https://github.com/columbia/pdslib/pull/95/commits/2258ed9001424942457528562e3a805f6c7ddcc7

File: src/pds/cross_report.rs (hunk 2 above)
Comment (giorgi-o) [reply src/pds/cross_report.rs]: Done in https://github.com/columbia/pdslib/pull/95/commits/925d709a57e6f5957bbe9b603604b1c2c1fa53e3

File: src/pds/cross_report.rs (hunk 1 above)
Comment (giorgi-o) [reply src/pds/cross_report.rs]: Changed `.insert(0.0)` to `.remove()` in https://github.com/columbia/pdslib/pull/95/commits/614c352e9e34d709433c1c2c40cd3e4886599477

File: src/pds/cross_report.rs (hunk 3 above)
Comment (giorgi-o) [reply src/pds/cross_report.rs]: Added in https://github.com/columbia/pdslib/pull/95/commits/614c352e9e34d709433c1c2c40cd3e4886599477

File: tests/ppa_demo.rs [hunk 8]
```diff
@@ -105,18 +102,14 @@ fn main() -> Result<(), anyhow::Error> {
             is_matching_event: Box::new(|event_filter_data: u64| {
                 event_filter_data == 1
             }),
-            bucket_intermediary_mapping: HashMap::new(),
+            requested_buckets: vec![0x559],
```
[the remark this answers is not in the document]
Comment (giorgi-o) [reply tests/ppa_demo.rs]: Added in https://github.com/columbia/pdslib/pull/95/commits/e6012cd5fe1679c6f6bdc0d78bcb6cbcda712f8d

**Comment by giorgi-o** [comment]
Your changes make much more sense, good ideas! LGTM
